<a href="https://colab.research.google.com/github/sulucay01/DI725-assignment1/blob/dev/notebooks/01_eda.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exploratory Data Analysis

## 1. Data Loading and Initial Inspection

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option("display.max_columns", None)
pd.set_option("display.max_colwidth", 200)

train = pd.read_csv("https://raw.githubusercontent.com/sulucay01/DI725-assignment1/main/data/raw/train.csv")
test = pd.read_csv("https://raw.githubusercontent.com/sulucay01/DI725-assignment1/main/data/raw/test.csv")

In [4]:
print("Train shape:", train.shape)
print("Test shape:", test.shape)

Train shape: (970, 11)
Test shape: (30, 11)


In [5]:
train.head()

,issue_area,issue_category,issue_sub_category,issue_category_sub_category,customer_sentiment,product_category,product_sub_category,issue_complexity,agent_experience_level,agent_experience_level_desc,conversation
0,Login and Account,Mobile Number and Email Verification,Verification requirement for mobile number or email address during login,Mobile Number and Email Verification -> Verification requirement for mobile number or email address during login,neutral,Appliances,Oven Toaster Grills (OTG),medium,junior,"handles customer inquiries independently, possess solid troubleshooting skills, and seek guidance from more experienced team members when needed.","Agent: Thank you for calling BrownBox Customer Support. My name is Tom. How may I assist you today?\n\nCustomer: Hi Tom, I'm trying to log in to my account to purchase an Oven Toaster Grill (OTG),..."
1,Cancellations and returns,Pickup and Shipping,Reasons for being asked to ship the item,Pickup and Shipping -> Reasons for being asked to ship the item,neutral,Electronics,Computer Monitor,less,junior,"handles customer inquiries independently, possess solid troubleshooting skills, and seek guidance from more experienced team members when needed.",Agent: Thank you for calling BrownBox customer support. My name is Alex. How may I assist you today?\n\nCustomer: Hi Alex. I recently received an email from BrownBox requesting me to ship back the...
2,Cancellations and returns,Replacement and Return Process,Inability to click the 'Cancel' button,Replacement and Return Process -> Inability to click the 'Cancel' button,neutral,Appliances,Juicer/Mixer/Grinder,medium,experienced,"confidently handles complex customer issues, excel in de-escalation, and possess the ability to empathize with customers, providing them with effective solutions and support.","Agent: Thank you for calling BrownBox Customer Support. My name is Sarah. How may I assist you today?\n\nCustomer: Hi Sarah, I am calling because I am unable to click the 'Cancel' button for my Ju..."
3,Login and Account,Login Issues and Error Messages,Error message regarding exceeded attempts to enter the correct verification code,Login Issues and Error Messages -> Error message regarding exceeded attempts to enter the correct verification code,neutral,Appliances,Water Purifier,less,inexperienced,"may struggle with ambiguous queries, rely on clarification from customers or guidance from senior team members, find it difficult to de-escalate tense situations, and may rely on predefined steps ...","Customer: Hi, I am facing an issue while logging into my account. I am getting an error message saying that I have exceeded the number of attempts to enter the correct verification code.\n\nAgent:..."
4,Order,Order Delivery Issues,Delivery not attempted again,Order Delivery Issues -> Delivery not attempted again,negative,Electronics,Bp Monitor,medium,experienced,"confidently handles complex customer issues, excel in de-escalation, and possess the ability to empathize with customers, providing them with effective solutions and support.","Agent: Thank you for contacting BrownBox customer support. My name is Sarah. How can I assist you today?\n\nCustomer: Hi Sarah, I have an issue with my order. I received a BP monitor, but the deli..."


In [6]:
train.columns.tolist()

['issue_area',
 'issue_category',
 'issue_sub_category',
 'issue_category_sub_category',
 'customer_sentiment',
 'product_category',
 'product_sub_category',
 'issue_complexity',
 'agent_experience_level',
 'agent_experience_level_desc',
 'conversation']

The training dataset contains 970 rows and 11 columns, while the test dataset contains 30 rows and 11 columns. The data includes several categorical variables related to issue type, product, and agent experience, along with the conversation text and the target variable customer_sentiment.

## 2. Dataset Structure and Data Quality

In [10]:
#Inspect Data Types
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 970 entries, 0 to 969
Data columns (total 11 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   issue_area                   970 non-null    object
 1   issue_category               970 non-null    object
 2   issue_sub_category           970 non-null    object
 3   issue_category_sub_category  970 non-null    object
 4   customer_sentiment           970 non-null    object
 5   product_category             970 non-null    object
 6   product_sub_category         970 non-null    object
 7   issue_complexity             970 non-null    object
 8   agent_experience_level       970 non-null    object
 9   agent_experience_level_desc  970 non-null    object
 10  conversation                 970 non-null    object
dtypes: object(11)
memory usage: 83.5+ KB


The dataset contains **970 observations and 11 columns**, and all features are stored as `object` data types. Most of the variables represent **categorical attributes** describing the issue type, product category, and agent experience level.

The `conversation` column contains **unstructured text data**, which will serve as the primary input for the sentiment classification model. The target variable `customer_sentiment` indicates the sentiment of the conversation and contains three possible classes.

In [12]:
#Check Missing Values
missing = train.isnull().sum().sort_values(ascending=False)
missing

,0
issue_area,0
issue_category,0
issue_sub_category,0
issue_category_sub_category,0
customer_sentiment,0
product_category,0
product_sub_category,0
issue_complexity,0
agent_experience_level,0
agent_experience_level_desc,0


The missing value analysis shows that **none of the features contain missing values**. All columns have 970 non-null observations.

This indicates that the dataset is **complete and clean**, and no missing value handling or imputation is required before further analysis or modeling.

In [13]:
#Check Duplicate Rows
train.duplicated().sum()

np.int64(0)

The dataset contains **no duplicate rows**, meaning each observation represents a unique customer interaction. Therefore, no duplicate removal is necessary at this stage.

In [14]:
#Inspect Unique Values of Categorical Features
categorical_cols = train.select_dtypes(include="object").columns
categorical_cols

Index(['issue_area', 'issue_category', 'issue_sub_category',
       'issue_category_sub_category', 'customer_sentiment', 'product_category',
       'product_sub_category', 'issue_complexity', 'agent_experience_level',
       'agent_experience_level_desc', 'conversation'],
      dtype='object')

In [15]:
for col in categorical_cols:
    print(col, train[col].nunique())

issue_area 6
issue_category 40
issue_sub_category 109
issue_category_sub_category 109
customer_sentiment 3
product_category 3
product_sub_category 50
issue_complexity 3
agent_experience_level 3
agent_experience_level_desc 3
conversation 968


The categorical features show varying levels of cardinality:

- `issue_area` contains **6 unique categories**
- `issue_category` contains **40 categories**
- `issue_sub_category` and `issue_category_sub_category` contain **109 categories**
- `product_category` contains **3 categories**
- `product_sub_category` contains **50 categories**
- `issue_complexity` contains **3 levels**
- `agent_experience_level` and `agent_experience_level_desc` contain **3 levels**
- `customer_sentiment` contains **3 classes**
- `conversation` contains **968 unique text entries**

The high number of unique values in the `conversation` column confirms that most conversations are distinct, which is expected for customer service interactions.

Features such as `issue_sub_category` and `product_sub_category` have relatively high cardinality, which may require careful handling during feature encoding if they are used as additional inputs to the model.